src/detectors/FOFFMTRK/OffMomentumReconstruction_factory.cc

```diff
@@ -0,0 +1,240 @@
+
+	if(m_readout.empty()){ m_log->error("READOUT IS EMPTY!"); return; }
+
+        auto id_spec = m_geoSvc->detector()->readout(m_readout).idSpec();
+        try {
+            id_dec = id_spec.decoder();
+            if (!m_sectorField.empty()) {
+                sector_idx = id_dec->index(m_sectorField);
+                m_log->info("Find sector field {}, index = {}", m_sectorField, sector_idx);
+            }
+            if (!m_layerField.empty()) {
+                layer_idx = id_dec->index(m_layerField);
+                m_log->info("Find layer field {}, index = {}", m_layerField, layer_idx);
+            }
+        } catch (...) {
+            m_log->error("Failed to load ID decoder for {}", m_readout);
+            throw JException("Failed to load ID decoder");
+        }
+
+
+	m_log->info("OMD Decoding complete...");
+
+        // local detector name has higher priority
+        if (!m_localDetElement.empty()) {
+            try {
+                local = detector->detector(m_localDetElement);
+                m_log->info("Local coordinate system from DetElement {}", m_localDetElement);
+            } catch (...) {
+                m_log->error("Failed to locate local coordinate system from DetElement {}", m_localDetElement);
+                throw JException("Failed to locate local coordinate system");
+            }
+            // or get from fields
+        } else {
+            std::vector<std::pair<std::string, int>> fields;
+            for (auto &f: u_localDetFields) {
+                fields.emplace_back(f, 0);
+            }
+            local_mask = id_spec.get_mask(fields);
+            // use all fields if nothing provided
+            if (fields.empty()) {
+                local_mask = ~0;
+            }
+            // info() << fmt::format("Local DetElement mask {:#064b} from fields [{}]", local_mask,
+            //                      fmt::join(fields, ", "))
+            //        << endmsg;
+        }
+
+        double det = aXOMD[0][0] * aXOMD[1][1] - aXOMD[0][1] * aXOMD[1][0];
+
+        if (det == 0) {
+            m_log->error("Reco matrix determinant = 0! Matrix cannot be inverted! Double-check matrix!");
+            throw JException("Reco matrix determinant = 0! Matrix cannot be inverted! Double-check matrix!");
+        }
+
+        aXOMDinv[0][0] = aXOMD[1][1] / det;
+        aXOMDinv[0][1] = -aXOMD[0][1] / det;
+        aXOMDinv[1][0] = -aXOMD[1][0] / det;
+        aXOMDinv[1][1] = aXOMD[0][0] / det;
+
+        det = aYOMD[0][0] * aYOMD[1][1] - aYOMD[0][1] * aYOMD[1][0];
+
+        if (det == 0) {
+            m_log->error("Reco matrix determinant = 0! Matrix cannot be inverted! Double-check matrix!");
+            throw JException("Reco matrix determinant = 0! Matrix cannot be inverted! Double-check matrix!");
+        }
+
+        aYOMDinv[0][0] = aYOMD[1][1] / det;
+        aYOMDinv[0][1] = -aYOMD[0][1] / det;
+        aYOMDinv[1][0] = -aYOMD[1][0] / det;
+        aYOMDinv[1][1] = aYOMD[0][0] / det;
```

Comment:
This stuff looks awfully familiar. Can't it share algorithm with RP code https://github.com/eic/EICrecon/pull/536 ?

Reply:
I'd suggest that splitting the matrix functionality off into an algorithm that is called for each factory is a good opportunity for simplification. I'll open an issue for this so it can be handled after this is merged. Does that sound ok, @veprbl? If so, just resolve this conversation.

Reply:
https://github.com/eic/EICrecon/issues/721

Reply:
I agree with this course, thank you.